### PR TITLE
Common Trees and Tree Edition Distance

### PR DESCRIPTION
**Pull request by Morriar**
This PR make some cleaning and renaming in `abstract_tree` to introduce `Tree`.

Unlike BinTree, a Tree can have more than two children so it can represent hierarchical data like a file and folder or a table of content.

Tree also provides the service `distance_from` used to compute the Tree Edition Distance between two trees.

Signed-off-by: Alexandre Terrasa <alexandre@moz-code.org>

### CODE REVIEW

File: lib/trees/abstract_tree.nit [hunk 1]
```diff
@@ -36,13 +36,19 @@ abstract class TreeMap[K: Comparable, E]
...
-class TreeNode[K: Comparable, E]
+# Abstract node structure used in `TreeMap` implementations
+#
+# Nodes are defined recursively each node (except the root one) pointing to a parent.
+# Nodes can be used to store data with the `AbstractTreeNode::value` attribute.
+#
+# Formal parameters:
+# * `K`: key type (a `Comparable` one so nodes can be sorted by their keys)
+# * `E`: value type (to store your data)
+abstract class AbstractTreeNode[K: Comparable, E]
```
Comment (privat) [inline lib/trees/abstract_tree.nit]: Je suis pas fan d'avoir le `Abstract` dans le non, je vois pas trop ce que ca apporte de plus.

File: lib/trees/tree.nit [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Common tree structure
+#
+# Hierarchical representation of data using recursive collections of nodes.
+#
+# This one differs from the `bintree` and `rbtree` implementation since it allows
+# more tha two children by node.
```
Comment (privat) [inline lib/trees/tree.nit]: than*

File: lib/trees/tree.nit [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+module tree
+
+import abstract_tree
+
+# Tree structure
+#
+# The tree contains a `root` node. If `root == null` then the tree is empty.
+# Each node can have `0..n` children.
```
Comment (privat) [inline lib/trees/tree.nit]: `n` is undefined. Are the children in a Sequence? a Collection?

File: lib/trees/tree.nit [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+# tree.add_node(nc, na)
+# tree.add_node(nd, nc)
+#
+# assert tree.to_formatted_string == """
+# a
+#  b
+#  c
+#   d
+# """
+# ~~~
```
Comment (privat) [inline lib/trees/tree.nit]: In the example I do not understand what the `Int` is for

File: lib/trees/tree.nit [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+	# 10:a
+	#  20:b
+	#   0:c
+	# """
+	# tree = new Tree[Int, String].from_string(str)
+	# assert tree.nodes[10].value == "a"
+	# assert tree.nodes[20].value == "b"
+	# assert tree.nodes[0].value == "c"
+	# ~~~
+	var nodes = new HashMap[K, N]
```
Comment (privat) [inline lib/trees/tree.nit]: protected? private?

File: lib/trees/tree.nit [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+	# ~~~
+	fun postorder: Array[N] do
+		var root = self.root
+		var nodes = new Array[N]
+		if root == null then return nodes
+		root.recursive_postorder nodes
+		return nodes
+	end
+
+	# TODO comment and test
```
Comment (privat) [inline lib/trees/tree.nit]: just do it

File: lib/trees/tree.nit [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+	# ~~~
+	#
+	# See:
+	# > *Kaizhong Zhang and Dennis Shasha.*
+	# > Simple fast algorithms for the editing distance between trees and
+	# > related problems.
+	# > SIAM Journal of Computing, 18:1245–1262, 1989.
+	fun distance_from(o: SELF): Int do
+		# Fill dynamic result map with zeros
+		var tdist = new Array[Array[Int]]
```
Comment (privat) [inline lib/trees/tree.nit]: One day we will have an efficient Matrix class or something, one day...

File: lib/trees/tree.nit [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+#
+# assert tree.to_formatted_string == """
+# a
+#  b
+#  c
+#   d
+# """
+# ~~~
+#
+# Use `show_dot` to display a graphic representation of a tree using graphiz.
```
Comment (R4PaSs) [inline lib/trees/tree.nit]: :s/graphiz/graphviz

File: lib/trees/tree.nit [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+		if ls.length > 1 then
+			return new TreeNode[Int, String](ls.first.to_i, ls.last)
+		end
+		return new TreeNode[Int, String](id, line.trim)
+	end
+
+	# Add a node to the tree
+	#
+	# If `parent == null` then node is set as the root node.
+	# Set `node.parent = parent` otherelse.
```
Comment (R4PaSs) [inline lib/trees/tree.nit]: :s/otherelse/otherwise

File: lib/trees/tree.nit [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+	# TODO comment and test
+	# Compute the Tree Edition Distance between `self` and `o`.
+	#
+	# Tree distance is computed based on the minimum sequence of edit operation
+	# to go from a tree to another.
+	# Edit operations are insert, remove and update.
+	# Each operation is associated to cost that can be modified with `insert_cost`,
+	# `remove_cost` and `update_cost`.
+	#
+	# Two trees are considered similar if their edit distance if equals to 0.
```
Comment (R4PaSs) [inline lib/trees/tree.nit]: :s/if equals/equals

**Comment by R4PaSs** [comment]
:shipit:

File: lib/trees/abstract_tree.nit (hunk 1 above)
Comment (Morriar) [inline lib/trees/abstract_tree.nit]: In this case I will need a better name idea :)

I renamed this one so I can use `TreeNode` for the kind that goes in a `Tree`. Maybe `TreeMapNode` for the abstract one?

File: lib/trees/tree.nit (hunk 5 above)
Comment (Morriar) [inline lib/trees/tree.nit]: This seems to be a useful service. Maybe just type it with `Map` to disable statically write services?

File: lib/trees/tree.nit (hunk 6 above)
Comment (Morriar) [inline lib/trees/tree.nit]: My bad, it's already done the line under this one :)

File: lib/trees/tree.nit (hunk 7 above)
Comment (Morriar) [inline lib/trees/tree.nit]: A simple Matrix class can be added to the stdlib. If you want an efficient one like what Matlab does, this require more work.

**Comment by privat** [comment]
Something is bothering me in the design of the new `TreeMap` design: it is the Map part.

In the previous tree implementations, each one used the tree structure as an algorithm mean to implements fast collections (fast maps in fact): fast addition of a element, fast retrieval, fast suppression. In fact, the organisations of nodes in the original trees was completely used internally and not exposed to the user.

In what is proposed now, its the user that controls the organisation of the nodes in the tree, and the map part seems to be no more used (in fact, it is implemented with a hashmap that is in complete opposition of what the other trees are doing).

Maybe I am completely out of touch about what is really happening here, but I think that the new code is a basically an overloading of the `tree` name.
I also think that a better proposal could be to add your concerns about tree distances in the existing module `ordered_tree` (or just as a standalone module of algorithms like what is done with coloring or perfect hashing)

**Pull request closed without merging**